lock-api/src/main/java/com/palantir/lock/watch/ClientLockWatchEventLogImpl.java

```diff
@@ -0,0 +1,220 @@
+        this.snapshotUpdater = snapshotUpdater;
+    }
+
+    @Override
+    public synchronized Optional<IdentifiedVersion> processUpdate(LockWatchStateUpdate update) {
+        final ProcessingVisitor visitor;
+        if (!latestVersion.isPresent() || !update.logId().equals(latestVersion.get().id())) {
+            visitor = new NewLeaderVisitor();
+        } else {
+            visitor = new ProcessingVisitor();
+        }
+        update.accept(visitor);
+        return latestVersion;
+    }
+
+    @Override
+    public synchronized void removeOldEntries(IdentifiedVersion earliestVersion) {
+        Set<Map.Entry<Long, LockWatchEvent>> eventsToBeRemoved =
+                eventMap.headMap(earliestVersion.version()).entrySet();
+        snapshotUpdater.processEvents(
+                eventsToBeRemoved.stream().map(Map.Entry::getValue).collect(Collectors.toList()));
+        eventsToBeRemoved.clear();
+    }
+
+    /**
+     * @param timestampToVersion mapping from timestamp to identified version from client-side event cache;
+     * @param version            latest version that the client knows about; should be before timestamps in the
+     *                           mapping;
+     * @return lock watch events that occurred from (exclusive) the provided version, up to (inclusive) the latest
+     * version in the timestamp to version map.
+     */
+    @Override
+    public synchronized TransactionsLockWatchEvents getEventsForTransactions(
+            Map<Long, IdentifiedVersion> timestampToVersion,
+            Optional<IdentifiedVersion> version) {
+        Optional<IdentifiedVersion> versionInclusive = version.map(this::createInclusiveVersion);
+        IdentifiedVersion toVersion = Collections.max(timestampToVersion.values(), IdentifiedVersion.comparator());
+        IdentifiedVersion currentVersion = getLatestVersionAndVerify(toVersion);
+
+        if (!versionInclusive.isPresent() || differentLeaderOrTooFarBehind(currentVersion, versionInclusive.get())) {
+            return TransactionsLockWatchEvents.failure(snapshotUpdater.getSnapshot(currentVersion));
+        }
+
+        if (eventMap.isEmpty()) {
+            return TransactionsLockWatchEvents.success(ImmutableList.of(), timestampToVersion);
+        }
+
+        IdentifiedVersion fromVersion = versionInclusive.get();
+
+        return TransactionsLockWatchEvents.success(
+                new ArrayList<>(eventMap.subMap(fromVersion.version(), true, toVersion.version(), true).values()),
+                timestampToVersion);
+    }
+
+    @Override
+    public synchronized Optional<Set<LockDescriptor>> getEventsBetweenVersions(
+            IdentifiedVersion startVersion,
+            IdentifiedVersion endVersion) {
+        IdentifiedVersion currentVersion = getLatestVersionAndVerify(endVersion);
+        IdentifiedVersion fromVersion = createInclusiveVersion(startVersion);
+
+        if (differentLeaderOrTooFarBehind(currentVersion, fromVersion)) {
```

Comment:
you should define this method (in as far as it has to be a different method) _in terms of_ the other one. The logic is basically the same. But I don't think should have its own return type, so should basically be the same method.